### DISCUSSION TITLE
How to pass data that is going to be used in a system to a plugin?

### DISCUSSION THREAD

**Discussion by IceSentry** [bevyengine/bevy#2517] [Q&A] [answered]
So, I want to make a plugin that needs to load a spritesheet specified by the user. At first I thought I could just add the path directly in the Plugin struct, but the plugin builder doesn't have, as far as I know, a way to use the asset server. So, I'm not sure what would be the recommended approach here. Should I just make a NameOfPluginSettings Resource and ask the user to insert that? Is there maybe a way to query for the Plugin struct?

**Comment by cart** [comment, marked as answer]
We currently recommend (and use internally) the "plugin settings as resource" pattern. For example, the official [LogPlugin](https://github.com/bevyengine/bevy/blob/main/crates/bevy_log/src/lib.rs#L65) looks like this:

```rust
struct LogPlugin;

pub struct LogSettings {
    pub filter: String,
    pub level: Level,
}

impl Plugin for LogPlugin {
    fn build(&self, app: &mut AppBuilder) {
        let settings = app.world_mut().get_resource_or_insert_with(LogSettings::default);
        // initialize plugin
    }
}
```

We encourage this pattern for a few reasons:

1. Plays nicely with plugin groups
2. Enables using ECS change detection to respond to respond to settings changes at runtime 
3. We ultimately might add "plugin lifecycles", which would remove the ability to directly construct Plugin values.
